File: intersection of two arrays2.py
#                      Hashmap    sorted: Two Pointers  Sorted: Binary search     
#============================================================================
# // Time Complexity : O(n+m)        O(m+n)                 O(mlon(n)) [best for larger values] 
# // Space Complexity : O(m)          O(1)                  O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
# 1. Binary search 
# for every element in small array do Binary search
# for each search increase the low if we have found the value

# 2. Two Pointer
# use pointers at head of both the lists check if we get the match otherwise move forward

# 3. Hashmap
# run for loop on small array and put all element and its counts in hashmap
# run for loop on bigger array and look in presense of that element in map and decrease its count if found

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n1 = len(nums1)
        n2 = len(nums2)
        if n1<n2: self.intersect(nums2, nums1) 
        # nums2 is smaller 
        # nums1 is larger
        nums1.sort()
        nums2.sort()
        res =[]
        low = 0
        high = n1
        index = 0
        
        #run for loop on small array
        for i in nums2:
            #biindex will get the index return from func
            bindex = self.binarySearch(nums1, index, high-1, i)
            if bindex != -1:
                res.append(i)
                index = bindex + 1
        return res
                
        
    
    def binarySearch(self, arr, low, high, target):
        
        while low <= high:
            mid = (low+high)//2
            
            if arr[mid] == target:
                # This is for repetative values in the arrays
                # if target is at oth index or arr[mid] is first targer then return
                if mid == low or arr[mid] > arr[mid-1]:
                    return mid
                # if false means there are repetative values in the array
                else:
                    high = mid - 1
                    
            elif arr[mid] < target:
                low = mid + 1
            else:
                high = mid - 1
                
        return -1
            
            
            
            
            
            
            
            
            
            
        
#         USING TWO POINTERS
#         while index<n2 and left<n1:
#             if nums1[left] == nums2[index]:
#                 res.append(nums1[left])
#                 index += 1
#                 left += 1
                
#             elif nums1[left] > nums2[index]:
#                 index += 1
            
#             elif nums1[left] < nums2[index]:
#                 left += 1
                
#         return res
        
        




        #        
#         for i in nums2:
#             if i not in hashmap:
#                 hashmap[i] = 0
#             hashmap[i] += 1
        
#         for j in nums1:
#             if j in hashmap:
#                 res.append(j)
#                 hashmap[j] -= 1
#                 if hashmap[j] == 0:
#                     del hashmap[j]
#         return res
            

File: Intersection_of_two_arrays2.java
//Hashmap solution
//TC:  O(n+m) where n is length of nums1 , m is length of nums2
//SC:O(min(n,m)) + O(li)  ,where li is arraylist used for storing elements 
//   hashmap space + list space

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        //null
        if(nums1.length ==0 || nums2.length ==0) return new int[0];
        if(nums1.length <nums2.length) return intersect(nums2,nums1);   //considering first array to be bigger one
        HashMap<Integer,Integer> map = new HashMap<>();
        
        for(int num : nums2){
            map.put(num , map.getOrDefault(num,0)+1);   //creating a hashmap of smaller array
        }
        
        List<Integer> li = new ArrayList<>();
        for(int num :nums1){     //checking if ele from nums1 exists in map
            if(map.containsKey(num)){
                li.add(num);   //if it exists then add it to list
                map.put(num,map.get(num)-1);  //decraese count of that ele in map
                map.remove(num,0);   //if value becomes 0 ,then remove it from map
            }
        }
        int[] result = new int[li.size()];
        for(int i=0;i<result.length;i++){
            result[i]=li.get(i);
        }
        return result;
    }
}

//Two pointer
//TC: O(m+n) ~ O(max(m,n))  where n is length of nums1 , m is length of nums2
//SC: O(li) -> arraylist space


class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        //null
        if(nums1.length ==0 || nums2.length ==0) return new int[0];
        if(nums1.length <nums2.length) return intersect(nums2,nums1);  //considering first array to be bigger one
       int n= nums1.length;
        int m = nums2.length;
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        
        List<Integer> li = new ArrayList<>();
        int  i=0,j=0;
        while(i<n && j<m){      
            if(nums1[i] == nums2[j]){   //if ele at both the ptr matched add it to list amd move ptrs
                li.add(nums1[i]);
                i++;j++;
            }else if(nums1[i]<nums2[j]){  //move the ptr with the smaller ele value
                i++;
            }else{
                j++;
            }
        }
        
        int[] result = new int[li.size()];
        for(int k=0;k<result.length;k++){
            result[k]=li.get(k);
        }
        return result;
    }
}

//Binary search
//TC: O(nlogm) where m is the length of bigger array
//SC: O(li) ;arraylist space

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        //null
        if(nums1.length ==0 || nums2.length ==0) return new int[0];
        if(nums1.length <nums2.length) return intersect(nums2,nums1);
       int n= nums1.length;
        int m = nums2.length;
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        int low =0;
        int high = n-1;
        
        List<Integer> li = new ArrayList<>();
        
        for(int i=0;i<m;i++){
            int bIndex = binarySearch(nums1,low , high , nums2[i]);  //search for the ele in the smaller array using binary search in bigger array
            if(bIndex != -1){
                low = bIndex+1;  //once an ele is found , move the low ptr
                li.add(nums2[i]);
            }
        }
       
        
        int[] result = new int[li.size()];
        for(int k=0;k<result.length;k++){
            result[k]=li.get(k);
        }
        return result;
    }
    
    private int binarySearch(int[] nums1 , int low , int high , int target){
        while(low<=high){
            int mid = low+(high-low)/2;
            if(nums1[mid]==target){
                //leftmost ele
                if(mid ==low|| nums1[mid-1]<nums1[mid] ){
                    return mid;
                }else{
                    high=mid-1;
                }
            }else if(nums1[mid]<target){
                low = mid+1;
            }else{
                high = mid-1;
            }
        }
        return -1;
    }
}